From 1e6736790cb09948ef2727a85ca4269948a1d89a Mon Sep 17 00:00:00 2001
From: Evgeny Voropaev <evorop@gmail.com>
Date: Fri, 4 Apr 2025 14:10:16 +0800
Subject: [PATCH v67 08/12] Fix page inconsistency on an empty page in
 heap_insert

Added a condition checking weither a page was really empty or not. The
condition was added into the logic deciding on the matter of INSERT+INT xlog
record type on the DO side.

Author: Evgeny Voropaev <evgeny.voropaev@tantorlabs.com>
Author: Maxim Orlov <orlovmg@gmail.com>
Reviewed-by: Yura Sokolov <funny.falcon@gmail.com>

Discussion:https://www.postgresql.org/message-id/4eb56320-744e-49ba-b766-702bc2fb61a8%40tantorlabs.com
---
 src/backend/access/heap/heapam.c | 20 ++++++++++++++++----
 1 file changed, 16 insertions(+), 4 deletions(-)

diff --git a/src/backend/access/heap/heapam.c b/src/backend/access/heap/heapam.c
index 86468a7d3e..cb1ee3684c 100644
--- a/src/backend/access/heap/heapam.c
+++ b/src/backend/access/heap/heapam.c
@@ -2149,6 +2149,9 @@ heap_insert(Relation relation, HeapTuple tup, CommandId cid,
 	Buffer		buffer;
 	Buffer		vmbuffer = InvalidBuffer;
 	bool		all_visible_cleared = false;
+	Page		page;
+	PageHeader	phdr;
+	bool		page_reinit_needed;
 
 	/* Cheap, simplistic check that the tuple matches the rel's rowtype. */
 	Assert(HeapTupleHeaderGetNatts(tup->t_data) <=
@@ -2196,13 +2199,22 @@ heap_insert(Relation relation, HeapTuple tup, CommandId cid,
 	/* NO EREPORT(ERROR) from here till changes are logged */
 	START_CRIT_SECTION();
 
+	/*
+	 * To decide whether or not to set the XLOG_HEAP_INIT_PAGE flag, we must
+	 * first determine if the page was physically empty before placeing a tuple.
+	 */
+	page = BufferGetPage(buffer);
+	phdr = (PageHeader) page;
+	page_reinit_needed = RelationNeedsWAL(relation) &&
+							(phdr->pd_special == phdr->pd_upper);
+
 	RelationPutHeapTuple(relation, buffer, heaptup,
 						 (options & HEAP_INSERT_SPECULATIVE) != 0);
 
-	if (PageIsAllVisible(BufferGetPage(buffer)))
+	if (PageIsAllVisible(page))
 	{
 		all_visible_cleared = true;
-		PageClearAllVisible(BufferGetPage(buffer));
+		PageClearAllVisible(page);
 		visibilitymap_clear(relation,
 							ItemPointerGetBlockNumber(&(heaptup->t_self)),
 							vmbuffer, VISIBILITYMAP_VALID_BITS);
@@ -2227,7 +2239,6 @@ heap_insert(Relation relation, HeapTuple tup, CommandId cid,
 		xl_heap_insert xlrec;
 		xl_heap_header xlhdr;
 		XLogRecPtr	recptr;
-		Page		page = BufferGetPage(buffer);
 		uint8		info = XLOG_HEAP_INSERT;
 		int			bufflags = 0;
 		TransactionId	xid_base,
@@ -2246,7 +2257,8 @@ heap_insert(Relation relation, HeapTuple tup, CommandId cid,
 		 * buffer references from XLogInsert.
 		 */
 		if (ItemPointerGetOffsetNumber(&(heaptup->t_self)) == FirstOffsetNumber &&
-			PageGetMaxOffsetNumber(page) == FirstOffsetNumber)
+			PageGetMaxOffsetNumber(page) == FirstOffsetNumber &&
+			page_reinit_needed)
 		{
 			info |= XLOG_HEAP_INIT_PAGE;
 			bufflags |= REGBUF_WILL_INIT;
-- 
2.51.0

